text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,902 @@
+- Feature Name: `const_trait_methods`
+- Start Date: 2024-12-13
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#67792](https://github.com/rust-lang/rust/issues/67792)
+
+# Summary
+[summary]: #summary
+
+Make trait methods callable in const contexts. This includes the following parts:
+
+* Allow marking `trait` declarations as const implementable.
+* Allow marking `trait` impls as `const`.
+* Allow marking trait bounds as `const` to make methods of them callable in const contexts.
+
+Fully contained example ([Playground of currently working example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2ab8d572c63bcf116b93c632705ddc1b)):
+
+```rust
+const trait Default {
+    fn default() -> Self;
+}
+
+impl const Default for () {
+    fn default() {}
+}
+
+const fn default<T: ~const Default>() -> T {
+    T::default()
+}
+
+fn compile_time_default<T: const Default>() -> T {
+    const { T::default() }
+}
+
+const _: () = Default::default();
+
+fn main() {
+    let () = default();
+    let () = compile_time_default();
+    let () = Default::default();
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+Const code is currently only able to use a small subset of Rust code, as many standard library APIs and builtin syntax things require calling trait methods to work.
+As an example, in const contexts you cannot use even basic equality on anything but primitives:
+
+```rust
+const fn foo() {
+    let a = [1, 2, 3];
+    let b = [1, 2, 4];
+    if a == b {} // ERROR: cannot call non-const operator in constant functions
+}
+```
+
+## Background
+
+This RFC requires familarity with "const contexts", so you may have to read [the relevant reference section](https://doc.rust-lang.org/reference/const_eval.html#const-context) first.
+
+Calling functions during const eval requires those functions' bodies to only use statements that const eval can handle. While it's possible to just run any code until it hits a statement const eval cannot handle, that would mean the function body is part of its semver guarantees. Something as innocent as a logging statement would make the function uncallable during const eval.
+
+Thus we have a marker (`const`) to add in front of functions that requires the function body to only contain things const eval can handle. This in turn allows a `const` annotated function to be called from const contexts, as you now have a guarantee it will stay callable.
```

Comment:
It is worth mentioning that we do defer *some* of these checks to const-eval time in a sense, such as determining whether `ptr.is_null()` can be executed. So our guarantee about things being const-callable is not 100% watertight. That's okay, it's not a load-bearing member for soundness, and we of course generally want to ensure const-evaluatability statically before we try to actually run the code, for the same reason that we prefer compilation errors over panics.

Reply:
The RFC as specified does not expose such situations, so I think we can ignore it for now.